// blapi/dist/bttps.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var node_fetch_1 = __importDefault(require("node-fetch"));
/** Custom post function based on node-fetch */
function post(apiPath, apiKey, sendObj, logStuff) {
    return __awaiter(this, void 0, void 0, function () {
        var postData, response, _a, _b, _c, e_1;
        return __generator(this, function (_d) {
            switch (_d.label) {
                case 0:
                    postData = JSON.stringify(sendObj);
                    _d.label = 1;
                case 1:
                    _d.trys.push([1, 5, , 6]);
                    return [4 /*yield*/, node_fetch_1.default(apiPath, {
                            method: 'POST',
                            body: postData,
                            headers: {
                                'Content-Type': 'application/json',
                                'Content-Length': String(postData.length),
                                Authorization: apiKey,
                            },
                        })];
                case 2:
                    response = _d.sent();
                    if (!logStuff) return [3 /*break*/, 4];
                    console.log('BLAPI: posted to', apiPath);
                    console.log('BLAPI: statusCode:', response.status, response.statusText);
                    console.log('BLAPI: headers:', response.headers.raw());
                    // it's text because text accepts both json and plain text, while json only supports json
                    _b = (_a = console).log;
                    _c = ['BLAPI: data:'];
                    return [4 /*yield*/, response.text()];
                case 3:
                    // it's text because text accepts both json and plain text, while json only supports json
                    _b.apply(_a, _c.concat([_d.sent()]));
                    _d.label = 4;
                case 4: return [2 /*return*/, response];
                case 5:
                    e_1 = _d.sent();
                    console.error('BLAPI:', e_1);
                    return [2 /*return*/, { error: e_1 }];
                case 6: return [2 /*return*/];
            }
        });
    });
}
exports.post = post;
/** Custom get function based on node-fetch */
function get(url) {
    return __awaiter(this, void 0, void 0, function () {
        var response, e_2;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    _a.trys.push([0, 2, , 3]);
                    return [4 /*yield*/, node_fetch_1.default(url)];
                case 1:
                    response = _a.sent();
                    return [2 /*return*/, response.json()];
                case 2:
                    e_2 = _a.sent();
                    console.error('BLAPI:', e_2);
                    throw new Error("Request to " + url + " failed with Errorcode " + e_2);
                case 3: return [2 /*return*/];
            }
        });
    });
}
exports.get = get;
